compiler/src/main/kotlin/io/spine/protodata/backend/AstComponents.kt

```diff
@@ -0,0 +1,197 @@
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Redistribution and use in source and/or binary forms, with or without
+ * modification, must retain the above copyright notice and the following
+ * disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.protodata.backend
+
+import com.google.protobuf.Descriptors.EnumValueDescriptor
+import com.google.protobuf.Descriptors.FieldDescriptor
+import com.google.protobuf.Descriptors.FileDescriptor
+import com.google.protobuf.Descriptors.MethodDescriptor
+import com.google.protobuf.Empty
+import io.spine.protodata.EnumConstant
+import io.spine.protodata.Field
+import io.spine.protodata.FieldKt
+import io.spine.protodata.FieldKt.ofMap
+import io.spine.protodata.File
+import io.spine.protodata.Rpc
+import io.spine.protodata.ServiceName
+import io.spine.protodata.TypeName
+import io.spine.protodata.cardinality
+import io.spine.protodata.constantName
+import io.spine.protodata.copy
+import io.spine.protodata.enumConstant
+import io.spine.protodata.field
+import io.spine.protodata.file
+import io.spine.protodata.name
+import io.spine.protodata.path
+import io.spine.protodata.rpc
+
+/**
+ * Converts this field descriptor into a [Field] with options.
+ *
+ * @see buildField
+ */
+internal fun FieldDescriptor.buildFieldWithOptions(
+    declaringType: TypeName,
+    documentation: Documentation
+): Field {
+    val field = buildField(declaringType, documentation)
+    return field.copy {
+        // There are several similar expressions in this file. Sadly, they have no common
+        // compile-time type which would allow us to extract the duplicates into a function.
+        option.addAll(listOptions(options))
+    }
+}
+
+/**
+ * Converts this field descriptor into a [Field].
+ *
+ * The resulting [Field] will not reflect the field options.
+ *
+ * @see buildFieldWithOptions
+ */
+internal fun FieldDescriptor.buildField(
+    declaringType: TypeName,
+    documentation: Documentation
+) = field {
+    name = name()
+    orderOfDeclaration = index
+    doc = documentation.forField(this@buildField)
+    this@field.number = this@buildField.number
+    this@field.declaringType = declaringType
+    copyTypeAndCardinality(this@buildField)
```

Comment:
Please extract `this@buildField` into a variable (e.g. `fieldDescr`, or `thisField`). This way you'd have fewer “dogs” around.